cmd/crank/beta/top/top.go

```diff
@@ -0,0 +1,164 @@
+limitations under the License.
+*/
+
+// Package top contains the top command.
+package top
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/alecthomas/kong"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/metrics/pkg/client/clientset/versioned"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+const (
+	errKubeConfig             = "failed to get kubeconfig"
+	errCreateK8sClientset     = "could not create the clientset for Kubernetes"
+	errCreateMetricsClientset = "could not create the clientset for Metrics"
+	errFetchAllPods           = "could not fetch all pods from all namespaces"
+	errGetPodMetrics          = "error getting metrics for pod"
+	errAddingPodMetrics       = "error adding metrics to pod"
+)
+
+// Cmd represents the top command.
+type Cmd struct{}
+
+type topMetrics struct {
+	podType      string
+	podName      string
+	podNamespace string
+	cpuUsage     int64
+	memoryUsage  string
+}
+
+// Help returns help instructions for the top command.
+func (c *Cmd) Help() string {
+	return `Usage: crossplane top`
+}
+
+// Run runs the top command.
+func (c *Cmd) Run(_ *kong.Context, logger logging.Logger) error {
+	logger = logger.WithValues("cmd", "top")
+
+	// Build the config from the kubeconfig path
+	config, err := ctrl.GetConfig()
+	if err != nil {
+		return errors.Wrap(err, errKubeConfig)
+	}
+	logger.Debug("Found kubeconfig")
+
+	// Create the clientset for Kubernetes
+	k8sClientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, errCreateK8sClientset)
+	}
+	logger.Debug("Created clientset for Kubernetes")
+
+	// Create the clientset for Metrics
+	metricsClientset, err := versioned.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, errCreateMetricsClientset)
+	}
+	logger.Debug("Created clientset for Metrics")
+
+	pods, err := k8sClientset.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, errFetchAllPods)
+	}
+
+	crossplanePods := c.getCrossplanePods(pods)
+	logger.Debug("Fetched all pods from all namespaces and filtered out Crossplane pods")
+
+	_, err = addPodMetrics(crossplanePods, metricsClientset)
+	if err != nil {
+		return errors.Wrap(err, errGetPodMetrics)
+	}
+	logger.Debug("Added metrics to Crossplane pods")
+
+	fmt.Println("TYPE                 NAMESPACE                 NAME                                                         CPU(cores)   MEMORY(bytes)")
+	for _, pod := range crossplanePods {
+		fmt.Printf("%-20s %-25s %-60s %-12d %-15s\n", pod.podType, pod.podNamespace, pod.podName, pod.cpuUsage, pod.memoryUsage)
+	}
+
+	return nil
+}
+
+func (c *Cmd) getCrossplanePods(pods *v1.PodList) []topMetrics {
```

Comment:
NIT: this does not need to be a receiver

Reply:
Good call, I was thinking about it too. It's more of a generic support function rather than command specific.